### PR TITLE
NRTs and Testing and Interface Changes, oh my.

### PR DESCRIPTION
**Pull request by msieker**
First, I did a dumb when starting work on these changes: I didn't make sure I had pulled the latest development. I believe I got all of the changes, but double checking those would be good. Gitkraken and github are both very confused by everything getting unindented moving to file scoped namespaces, and I ended up pulling out Winmerge to check everything. Is there an .editorconfig or a R# config for some of the formatting/styles?

Now, for actual changes. These mainly stemmed from how I'm exporting with SimpleFrat and an issue I ran into when wiring up the last bits for that. That resulted in this mess of changes. 

## In retrospect, a false start.

The inciting incident resulted from this line in `SpreadsheetConfiguration.cs`:
```c#
public Type DataType => typeof(T);
```
and how `ISpreadsheetGenerator.cs` defined this function:
```c#
byte[] CreateMultiSheetSpreadsheet(IEnumerable<ISpreadsheetConfiguration<object>> exportSheets)
```
This had the end result of the type discover deciding everything is an object and not actually exporting anything.

And at this point, I have egg on my face, I swear earlier tonight before doing anything, I tried this:
```c#
var config = new ISpreadsheetConfiguration<object>[]
{
	new SpreadsheetConfiguration<SampleExportRecord>(),
	new SpreadsheetConfiguration<PersonRecord>(),
};
```
and had the compiler laugh at me. Turns out what I tried was this:
```c#
var config = new SpreadsheetConfiguration<object>[]
{
	new SpreadsheetConfiguration<SampleExportRecord>(),
	new SpreadsheetConfiguration<PersonRecord>(),
};
```
Which is a mostly reasonable thing to try. However, I've tackled this from two ways that I think is an improvement to using the library.

The first is splitting up the `ISpreadsheetConfiguration` interface. It now has a non-generic version where `ExportData` is an `IEnumberable`, along with a generic version that uses `new` and explicit implementation to make an `ExportData` that is `IEnumberable<T>`.  `ISpreadsheetGenerator.CreateMultiSheetSpreadsheet` now uses the non-generic version. This drops the `<object>` wart from a bunch of functions, and makes it easier to choose the right types. 

The second is something that just sweeps all that under the rug and just does the right thing on its own: The `MultisheetConfiguration`:
```c#
var config = new MultisheetConfiguration()
	.WithSheet("Sheet 1", testSheet1Data)
	.WithSheet("Sheet 2", testSheet2Data, conf =>
	{
		conf.AutoSizeColumns = false;
	});
```
Overloads have been added to `CreateMultiSheetSpreadsheet` to take this object. All it is is a class implementing `IEnumerable<ISpreadsheetConfiguration>` and delegating it to an internal list. The overloads just cast it to the `IEnumerable` and pass it on.

## NRT and Nullability
Both the parser and generator are aware of nullable now, and will do the right thing. I've also turned on NRT in most files I've touched, which resulted in some additional checks for null in the import to shut the warnings up, along with a nice `SpreadsheetParserException` that can be thrown at those points. There was somewhere throwing a plain `Exception` that was changed to this type.

As part of this, the parser picked up knowledge of all the same types the generator does. It will also now throw a `SpreadsheetParserException` with a helpful message if you try to access a sheet number beyond what the workbook has. This was one of the added null checks, and figured it might as well be helpful instead of throwing an out of bounds exception.

## A Few Random Things
- Overloads have been added to the various `Create` methods in the exporter that take a `Stream` as a parameter, and return a bool instead of a byte array. The existing versions still exist, and just create a `MemoryStream` themselves to call the overloads.
- The parser had a `using (fileStream)` statement that would dispose of a stream that it didn't own. This has been removed. It's the caller's stream, it's their job to dispose it when they want to.

## Tests
These now cover a lot more than they did. It now uses a type for export that has both nullable and normal versions of all the types it knows about (Anything the exporter doesn't know about just gets `ToString`'d). Tests have been added for multisheet exporting. Additionally, the parser is now used for the generator tests, ensuring the files written can be parsed, and making sure that all of the expected values make it out to the spreadsheet. (incidentally, this was how the `using` issue in the parser was found, multisheet tests call the validation function multiple times with the same stream, breaking when it tried to seek a disposed stream).

## The One Time Excel Doesn't Think Something Is A Date
With all these changes done, I had covered all I could without actually trying to open a file in Excel. Upon doing so, I was greeted by a message saying the workbook was corrupt. After a bit of looking trying to figure out what it might have an issue with, I noticed that `DateTime`s were being displayed as numbers. `DateTimeOffset`s were fine. After a bit of digging, I discovered I had to write it out in OLE Automation format (Floating point number of days since 1900-1-1, with the decimal part the fraction of the day that has passed) with a custom style. The parser now also knows what to do if it encounters a floating point value when it expects something resembling a `DateTime`. After fiddling with the cell data type a bit, Excel finally opened the files without complaint.

### CODE REVIEW

File: src/NetCore.Utilities.Spreadsheet.Tests/OpenXmlSpreadsheetGeneratorTests.cs [hunk 1]
```diff
@@ -109,28 +110,178 @@ public void CreateMultiSheetSpreadsheet_ShouldThrowError_WhenProvidedSheetsIsNul
...
+                    ExportData = testSheet2Data
+                }
+        };
+        using var ms = new MemoryStream();
+        var result = _spreadsheetGenerator.CreateMultiSheetSpreadsheet(ms, config);
+
+        result.Should().BeTrue();
+        ms.Should().NotHaveLength(0);
+
+        //var sheetPath = Path.Join(Path.GetTempPath(), $"CreateMultiSheetSpreadsheet_With_A_Stream_Should_Work_{DateTime.Now:yyyyMMddHHmmssfff}.xlsx");
```
Comment (mitchelsellers) [inline src/NetCore.Utilities.Spreadsheet.Tests/OpenXmlSpreadsheetGeneratorTests.cs:168]: Lets remove the commented out stuff

File: src/NetCore.Utilities.Spreadsheet.Tests/OpenXmlSpreadsheetGeneratorTests.cs [hunk 2]
```diff
@@ -109,28 +110,178 @@ public void CreateMultiSheetSpreadsheet_ShouldThrowError_WhenProvidedSheetsIsNul
...
+        
+        var config = new MultisheetConfiguration()
+            .WithSheet("Sheet 1", testSheet1Data)
+            .WithSheet("Sheet 2", testSheet2Data, conf =>
+            {
+                conf.AutoSizeColumns = false;
+            });
+        var result = _spreadsheetGenerator.CreateMultiSheetSpreadsheet(config);
+        result.Should().NotBeNullOrEmpty();
+        //var sheetPath = Path.Join(Path.GetTempPath(), $"CreateMultiSheetSpreadsheet_With_A_Stream_Should_Work_{DateTime.Now:yyyyMMddHHmmssfff}.xlsx");
```
Comment (mitchelsellers) [inline src/NetCore.Utilities.Spreadsheet.Tests/OpenXmlSpreadsheetGeneratorTests.cs:193]: Same here

File: src/NetCore.Utilities.Spreadsheet/OpenXmlSpreadsheetGenerator.cs [hunk 3]
```diff
@@ -1,434 +1,453 @@
...
-            if (exportConfiguration.RenderTitle && string.IsNullOrEmpty(exportConfiguration.DocumentTitle))
-                throw new ArgumentException("Document Title is required when 'Render Title' is true",
-                    nameof(exportConfiguration.DocumentTitle));
+        var workbookPart = spreadsheetDocument.AddWorkbookPart();
+        workbookPart.Workbook = new Workbook();
 
-            if (exportConfiguration.RenderSubTitle && string.IsNullOrEmpty(exportConfiguration.DocumentSubTitle))
-                throw new ArgumentException("Document Sub Title is required when 'Render Sub Title' is true",
-                    nameof(exportConfiguration.DocumentSubTitle));
+        Debug.Assert(spreadsheetDocument.WorkbookPart != null, "spreadsheetDocument.WorkbookPart != null. Something must be wrong with the universe.");
```
Comment (mitchelsellers) [inline src/NetCore.Utilities.Spreadsheet/OpenXmlSpreadsheetGenerator.cs:54]: Lets remove this debug assertion

File: src/NetCore.Utilities.Spreadsheet/SpreadsheetConfiguration.cs [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 using System.Collections.Generic;
 
 namespace ICG.NetCore.Utilities.Spreadsheet;
 
 /// <summary>
 ///     Definition of a spreadsheet configuration for exporting of data
 /// </summary>
-/// <typeparam name="T">The type to be exported</typeparam>
-public interface ISpreadsheetConfiguration<out T> where T : class
+public interface ISpreadsheetConfiguration
```
Comment (mitchelsellers) [inline src/NetCore.Utilities.Spreadsheet/SpreadsheetConfiguration.cs:10]: I don't believe that this change is binary compatible and will introduce breaking changes, can we validate this behavior

**Comment by mitchelsellers** [comment]
@msieker Overall this is looking pretty good, and some nice improvements, but I want to validate a few things.

1. Binary compatibility with prior versions due to the changes to Multi Spreadsheet Exporting - IE do we need to increment version number by a major version
2. In the current usage, if you had a DateTime value and exported with a column format of "d" it worked as a proper export, if that was no-longer working I'd like to know why.  Was this possibly due to the changes necessary for support of Nullable

The removal of the usage of `<object>` to get to your more streamlined process seems to be a great contribution as well.  The current pattern was a great example of the usage of [Covariance](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/), but it isn't very commonly used it seems.

**Comment by mitchelsellers** [comment]
@kenny-sellers Can you review this one as well?

**Comment by msieker** [comment]
@mitchelsellers I wouldn't have a problem with the covariance if it was something other than object, having `<object>` as a generic parameter never seems right to me. Plus the only thing that generic parameter does is basically set the `DataType` property. If I wanted to go all out, I'd probably drop the overloads taking `IEnumerable<ISpreadsheetConfiguration>` from the public interface entirely, and just have the `MultisheetConfiguration` ones, making there be one (correct) way of using it.

Now as for your points:
1. This probably does break binary compatibility, since `ISpreadsheetGenerator` no longer has the `IEnumerable<ISpreadsheetConfiguration<object>>` overloads. However, as it is now, I don't believe it breaks source compatibility, the compiler just chooses different overloads. The example works without changes, and if you point me at an existing project or two that uses this, I can check those.
2. After some playing, I now have Excel convinced to properly handle DateTimes without jumping through the OADate hoop. The issue that broke dates was probably from my last round of changes. I didn't make tests as thorough as the ones I have now where I hammer on every type. I should not mess with Excel date formatting at 3am.

And poking with the formatting, I found a bug in `TypeDiscoverer` that had both the width and format variables in the wrong scope, causing all properties after the one it was specified on to take on that format and/or width. That's fixed. While I was there, I also fixed #30, since that was simply adding `.ToLowerInvariant()` when pulling the value out of the attribute.

**Comment by mitchelsellers** [comment]
Ok, I agree, I think overall this is much better,

Can you increment the GitVersion file major version number?  I think this is worthy of doing just in case, then I can approve.

**Comment by msieker** [comment]
Before doing that, should I do #29? Seems low effort, add three number styles.

There's also a few attributes that were marked as being removed in the next major version that should probably be cleaned up.

And do you prefer updating the `gitversion.yml` file compared to just doing `+semver: major` in the merge message per: [the GitVersion docs](https://gitversion.net/docs/reference/version-increments#commit-messages)?

**Comment by mitchelsellers** [comment]
@msieker Sure if you want to do #29 as well that works.  We could just add the semver commit message here as well that works.

I would agree, removing the old attributes would be proper

**Comment by msieker** [comment]
Attribute removed, fixed number styles added (along with dedicated columns in the output spreadsheet for validation):
image
I think this is ready to be buttoned up now.

**Review by mitchelsellers — APPROVED** [review]

**Pull request merged**